src/tasks.ts

```diff
@@ -0,0 +1,331 @@
+import { PubkeyHex } from '@lodestar/api/keymanager';
+import { Epoch, RootHex, Slot } from '@lodestar/types';
+import { SignedBeaconBlock } from '@lodestar/types/allForks';
+import { BlockTag } from 'ethers';
+import * as R from 'ramda';
+
+import { readArtifact, saveArtifact } from './artifacts.js';
+import * as Cache from './cache.js';
+import { catCID, uploadTree } from './ipfs.js';
+import { buildTree, loadTree } from './merkle.js';
+import { shared as Shared } from './shared.js';
+import { ValidatorInfo } from './types.js';
+import { debug, toHex } from './utils.js';
+
+export async function main(): Promise<void> {
+    // TODO: place somewhere else?
+    Cache.destroy();
+
+    const clBlock = await getLastFinalizedBeaconBlock();
+    const blockTag = getBlockTag(clBlock);
+
+    if (!(await isReportable(blockTag))) {
+        debug('Report is not allowed');
+        return;
+    }
+
+    const targetEpoch = await Shared.ORACLE.nextReportEpoch({
+        blockTag,
+    });
+    debug('Target epoch is', targetEpoch);
+
+    if (targetEpoch > slotToEpoch(clBlock.message.slot)) {
+        debug('Target epoch is not reached yet');
+        return;
+    }
+
+    // TODO: query report frame instead
+    const sourceEpoch = await Shared.ORACLE.lastConsolidatedEpoch({
+        blockTag,
+    });
+    debug('Source epoch is', sourceEpoch);
+
+    const artifact = readArtifact(Number(sourceEpoch), Number(targetEpoch));
+    if (!R.isNil(artifact)) {
+        debug('Report tree artifact already exists');
+        return;
+    }
+
+    const prevTreeCID = await Shared.ORACLE.treeCid({
+        blockTag,
+    });
+
+    if (R.isEmpty(prevTreeCID)) {
+        debug('No previous report tree CID found');
+    } else {
+        debug('Reading the previous tree from IPFS by CID', prevTreeCID);
+        const prevTree = loadTree(await catCID(prevTreeCID));
+
+        debug('Check the tree root');
+        const prevRoot = await Shared.ORACLE.reportRoot({
+            blockTag,
+        });
+
+        if (prevTree.root !== prevRoot) {
+            throw new Error('Tree root mismatch');
+        }
+
+        debug('Fill in the rewards cache');
+        for (const [_, [no, reward]] of prevTree.entries()) {
+            Cache.rewards.set(no, reward);
+        }
+    }
+
+    debug('Loading validators');
+    await loadValidators(Number(targetEpoch));
+
+    debug('Fetching duties');
+    // TODO: fix range
+    for (const epoch of R.range(Number(sourceEpoch), Number(targetEpoch))) {
```

Comment:
We need to think about the algorithm. The state queries is very hard for nodes (especially historical). I would think about possibility to keep this daemon works always, check new finalized epochs one by one (the nearest epochs is quite easy managed by nodes), write the results of every epoch to some file (so that no data is lost if the container is restarted, and the data of `225 epochs * N days` might be too large for the memory location) and when the time is come, build the report for the frame

Reply:
Our oracles run their own nodes alongside the oracle software as a requirement. I will check it against the similar setup, but I think it won't be a problem. Otherwise, we will take into account these considerations. The problem with this caching layer is that you have to implement a part of a node software and make it really robust.